packages/drivers/odsp-driver/src/snapshotFormatConverter.ts

```diff
@@ -0,0 +1,279 @@
+ * Licensed under the MIT License.
+ */
+
+import { assert, stringToBuffer } from "@fluidframework/common-utils";
+import { IBlob, ISnapshotTree } from "@fluidframework/protocol-definitions";
+import { ISequencedDeltaOpMessage } from "./contracts";
+import { BlobCore, NodeCore, ReadBuffer, TreeBuilder } from "./snapshotRepresentation";
+
+const minReadVersion = "1.0";
+
+/**
+ *  Header of the downloaded snapshot.
+ */
+export interface ISnapshotHeader {
+    // This is the minimum version of the reader required to read the wire format of a given snapshot.
+    minReadVersion: string,
+    // Represents the version with which snapshot is created.
+    createVersion: string,
+    // Seq number at which the snapshot is created.
+    snapshotSeqNumber: number,
+}
+
+/**
+ * Class used to convert snapshot formats. It can convert binary compact representation to
+ * odsp snapshot and vice-versa.
+ */
+export class SnapshotFormatConverter {
+    constructor() {}
+
+    /**
+     * Recreates header section and then validate that.
+     * @param node - tree node to read header section from
+     */
+    private readAndValidateHeaderSection(node: NodeCore) {
+        assert(node.getString(0) === "MinReadVersion", "MinReadVersion header should be present");
+        assert(node.getString(2) === "CreateVersion", "CreateVersion header should be present");
+        assert(node.getString(4) === "SnapshotSequenceNumber", "SnapshotSequenceNumber header should be present");
+        const header = {
+            minReadVersion: node.getString(1),
+            createVersion: node.getString(3),
+            snapshotSeqNumber: node.getNumber(5),
+        };
+        assert(minReadVersion >= header.minReadVersion, "Driver min read version should >= to server minReadVersion");
+        assert(header.createVersion >= minReadVersion, "Snapshot should be created with minReadVersion or above");
+    }
+
+    /**
+     * Recreates dictionary section of the tree.
+     * @param node - tree node to read dictionary section from. This container paths/ids mapping
+     * to integer representation.
+     */
+    private readDictionarySection(node: NodeCore) {
+        const dictionary = new Array<string>();
+        for (const name of node) {
+            assert(name instanceof BlobCore, "Mapping should be of type BlobCore");
+            dictionary.push(name.toString());
+        }
+        return dictionary;
+    }
+
+    /**
+     * Recreates blobs section of the tree.
+     * @param node - tree node to read blob section from
+     * @param dictionary - name map, used to decode path/IDs.
+     */
+    private readBlobSection(node: NodeCore, dictionary: string[]) {
+        const blobs: Map<string, Uint8Array> = new Map();
+        for (const [idIndex, blob] of node.iteratePairs()) {
+            assert(typeof idIndex === "number", "Blob index should be a number");
+            assert(blob instanceof BlobCore, "Blob content should be of type blob");
```

Comment:
We should think how to make these asserts actionable. I.e. if it is hit in production the way it's written, it's not really actionable. I.e. we need to convert them to exceptions with enough metadata that gets logged into telemetry, such that we can make some progress using it (or can keep adding metadata until it's clear where the problem is)

Reply:
I think we can defer this work until the actual format is implemented since we don't know these would remain or not.